### ISSUE TITLE
use of  OPENSHIFT_DATA_DIR

### ISSUE THREAD

**Issue by sonalkr132** [glittergallery/GlitterGallery#260]
#228 leads to broken image
![screenshot from 2015-03-13 09 56 22](https://cloud.githubusercontent.com/assets/7680662/6643228/a21de5ac-c9d0-11e4-8874-2850dff96560.png)
You might say, lets use 
```Ruby
elsif Rails.env.production?
	Glitter::Application.config.repo_dir=ENV['OPENSHIFT_DATA_DIR'] +"data/"
```
But things don't get any easier, as openshift seem to have bug related to OPENSHIFT_DATA_DIR. image's path get resolved to "/var/lib/openshift/xxxxx/app-root/data/....." instead of "app-root/data/..." where files are originally stored and where OPENSHIFT_DATA_DIR originally points to. look [here] (https://developers.openshift.com/en/managing-filesystem.html#__code_data_code). If interested further [read](https://bugzilla.redhat.com/show_bug.cgi?id=962432)

I kinda hacked my way with by adding following in .openshift/action_hooks/build:
`ln -sf ../../../../app-root/data ${OPENSHIFT_REPO_DIR}/public/data`

does anyone have better idea?

**Comment by Alwahsh** [comment]
In my opinion, the best thing to do is to go back to what we had. It works well and easier in setup.

**Comment by sonalkr132** [comment]
Didn't that have issue with having to delete "public/data" symlink file manually in deb env? 

My current implementation doesn't require that, and content uploaded by user still persists between deployments. I have added that line to build, so a soft link gets created every time we deploy, so kinda like what we had in beginning, but now that file is added during deployment(no need to worry about in dev)

What I was asking if someone has better idea than my implementation, or may be I was doing something wrong with use of "OPENSHIFT_DATA_DIR", which others can figure out. I have achieved the same effect but I hadn't used the "OPENSHIFT_DATA_DIR" var.

**Comment by Alwahsh** [comment]
Oh... I didn't read that line of adding it to the action_hook. Seems good to me except that I have a single suggestion: 
use this line instead of writing the path yourself:
` ln -sf ${OPENSHIFT_DATA_DIR} ${OPENSHIFT_REPO_DIR}/public/data`
I suggest you make a PR with those modifications.

**Comment by sonalkr132** [comment]
Thats the apparent bug I was talking about. During upload of content everything is fine(ie content is stored in app-root/data) but while fetching "OPENSHIFT_DATA_DIR" gets resolved to  "/var/lib/openshift/xxxxx/app-root/data/....." instead of "app-root/data/...". So i used the explicit path.

**Comment by Alwahsh** [comment]
I don't see it a problem that OPENSHIFT_DATA_DIR gets resolved to /var/lib/openshift/xxxxx/app-root/data/
It just means that it was wrong to use it to link to something in html but creating a symlink from the public folder to where it really is is the way to resolve it.
Please let me know if I'm missing something here :)

**Comment by sonalkr132** [comment]
OPENSHIFT_DATA_DIR resolving to "/var/lib/openshift/xxxxx/app-root/data/some.file" means that we sending request which looks like "http://new-glitterfollowers.rhcloud.com/var/lib/openshift/xxxxx/app-root/data/some.file" when file doesn't exist there. It exists at "http://new-glitterfollowers.rhcloud.com/app-root/data/some.file".

We can use absolute path instead of relative path? idk! I think at some point I tired "~app-root/data", I don't remember it working. Besides, it's a PaaS not IaaS, so we don't have access to everything.

**Comment by Alwahsh** [comment]
OK... I understand that... I'm not sure though how my suggestion hits that issue... This issue will appear if we directly link to the file from within the rails app using `OPENSHIFT_DATA_DIR` but we can use it safely for the symlink...

**Comment by sonalkr132** [comment]
nope.. we can't. I tried that too -_- thats one of the reason it took me like a day to get everything working on openshift. 

```Ruby
elsif Rails.env.production?
  Glitter::Application.config.repo_dir=ENV['OPENSHIFT_DATA_DIR'] + "data/"
```
and 
```Ruby
ln -sf ${OPENSHIFT_DATA_DIR} ${OPENSHIFT_REPO_DIR}/public/data
```
have same effect. they both render pictures with src = "var/lib/openshift/xxxxx/app-root/data/...." when it should use "app-root/data/...."

**Comment by Alwahsh** [comment]
The first piece of code is definitely wrong.
The second one however... I'm kinda sure works if you fix the first one, I can't imagine that on a Rails app you'd say 
`link_to "link","data/something"` would be translated into "var/lib/.." because data is a symlink to that... It really makes no sense :)

**Comment by sonalkr132** [comment]
we have "${OPENSHIFT_REPO_DIR}/public/data" symlink to "${OPENSHIFT_DATA_DIR}" and ""${OPENSHIFT_DATA_DIR}" resolves to "var/lib/openshift/xxxxx/app-root/data/....". So it's like

${OPENSHIFT_REPO_DIR}/public/data -> ${OPENSHIFT_DATA_DIR} -> var/lib/openshift/xxxxx/app-root/data/....

But during upload using same "${OPENSHIFT_DATA_DIR}" content get stored at "app-root/data/...." weird  right? lol. try it yourself if you don't believe me.

**Comment by Alwahsh** [comment]
I'd say that the behavior you described is what makes sense to me. That's what I expect to happen at least :)

I believe what happenned is as follows:
when you tried the line I suggested, you forgot to modify this part:
`Glitter::Application.config.repo_dir=ENV['OPENSHIFT_DATA_DIR'] +"data/"`
which would have the effect of making upload only works so you thought it was because of the `OPENSHIFT_DATA_DIR`

**Comment by sonalkr132** [comment]
nope! I was using
```Ruby
Glitter::Application.config.repo_dir = "public/data"
```
which would hit the symlink created by
```
ln -sf ${OPENSHIFT_DATA_DIR} ${OPENSHIFT_REPO_DIR}/public/data
```
Seriouly! try it yourself if you find it hard to believe.

**Comment by Alwahsh** [comment]
It works as expected :)

**Comment by sonalkr132** [comment]
let me try again. I remember it precisely that i tried your way too.

**Comment by Alwahsh** [comment]
I expect that it has all gone as I told you before. You simply forgot to modify the code :)

**Comment by sonalkr132** [comment]
damn! it works. 

Sorry.. my bad!

**Comment by Alwahsh** [comment]
np, it's cool :)

**Comment by sonalkr132** [comment]
Do you think adding an
```
if
....
fi

would make sense to check if file already exiits? Most people used it, but in deployment everything gets erased from app-deployment, so would be that symlink.

**Comment by rohitpaulk** [comment]
> If you use a scalable application and you benefit from OpenShift’s scaling features, this directory is not shared among gears. In such case, you should use something like S3 and outsource your storage.

> it is not safe to write files anywhere else except this directory and only in non-scalable mode.

We're sticking to OpenShift now, so run your demos in non-scalable mode and we'll be fine. This is a limitation of OpenShift, I don't think that we should waste our time working around it.

**Comment by sonalkr132** [comment]
"this directory" as in? app-root/data?

**Comment by Alwahsh** [comment]
@rohitpaulk you might want to check [that line](https://github.com/glittergallery/GlitterGallery/blob/master/app/models/project.rb#L47):
See if we don't define it as "public/data" you're going to link to it at the HTML level incorrectly.

**Comment by rohitpaulk** [comment]
@Alwahsh - You're right. My solution wouldn't work unless we're serving the assets from `OPENSHIFT_DATA_DIR`. I'm going to merge #264.

**Comment by Alwahsh** [comment]
@rohitpaulk I also think that since we're forcing the configuration to be a certain thing, wouldn't it be better if we disallow this configuration at all?

**Comment by rohitpaulk** [comment]
@Alwahsh - from what I understand - As long as the directory is within `public`, we're fine. The reason I introduced the config there initially, was for the test directory. Our tests use `/public/testdata` or something similar - it's washed out after the tests are done. 

The proper solution to this would be to **not** force the configuration to be a certain thing. We should support interchangeable storage mechanisms. Local (within `public/`) and S3 would be a good start. OpenShift's persistent storage is a special case, I guess. (where we have to use their build hooks and not our app code).

**Comment by sonalkr132** [comment]
@rohitpaulk 
>If you use a scalable application and you benefit from OpenShift’s scaling features, this directory is >not shared among gears. In such case, you should use something like S3 and outsource your storage.

>it is not safe to write files anywhere else except this directory and only in non-scalable mode.

This is true for scaleable apps, but by default apps are non-scaleable. [link](https://developers.openshift.com/en/managing-scaling.html#creating-a-scalable-application) So demos we are running are non-scaleable by default.

And can someone be kind enough to explain what is meant by "disallowing configuration" ? I have no clue what it is :confused:

**Comment by Alwahsh** [comment]
@sonalkr132 I meant to not allow users to set the folder they want to use for data by themselves as in [here](https://github.com/glittergallery/GlitterGallery/blob/master/config/initializers/0_glitterconfig.rb#L42) but now I believe it is better to allow them full control... I already have in mind some setup to allow for full configuration.

**Comment by sonalkr132** [comment]
What advantage would we have if we let someone decide where does he want the repos to be store?

**Comment by Alwahsh** [comment]
it's not to our advantage but the user would definitely have advantages... Think of this, if we didn't allow him to configure where he wants the repos to be stored, we might have used a path that gets deleted by openshift on deployment which would be bad. There's really countless advantages to the user from being able to decide :)

**Comment by EmilyDirsh** [comment]
We should keep the configuration. We're running demos on OpenShift, but we can't assume that other people might not want to run GG in other environments, and while we don't have to specifically support that, we shouldn't go out of our way to hinder it, either.
